torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp

```diff
@@ -0,0 +1,916 @@
+    SegmentedGroup* group) {
+  Fusion* fusion = &segmented_fusion->completeFusion();
+  auto h = tryMerge(fusion, group);
+  TORCH_INTERNAL_ASSERT(h.has_value());
+  return h.value();
+}
+
+SegmentCandidateFinder::SegmentCandidateFinder(const Fusion* fusion) {
+  segmented_fusion = std::make_unique<SegmentedFusion>(fusion);
+  findSegments();
+}
+
+void SegmentCandidateFinder::findSegments() {
+  // TODO: Make traversal items local to this function.
+
+  // Need this for initialization of the DAG that is process
+  std::unordered_map<Expr*, SegmentedGroup*> expr2group;
+
+  // Initialize DAG, convert each expr to a segment group
+  size_t total_exprs = 0;
+  for (auto expr : completeFusion().exprs()) {
+    auto new_group = segmented_fusion->newGroup(expr);
+    expr2group.insert(std::make_pair(expr, new_group));
+    total_exprs++;
+  }
+
+  segmented_fusion->total_expr_count_ = total_exprs;
+
+  // Create edges between the Exprs. Mark inputs and outputs of the fusion.
+  for (auto expr : completeFusion().exprs()) {
+    auto expr_group = expr2group.at(expr);
+    for (auto inp : expr->inputs()) {
+      if (inp->isFusionInput()) {
+        expr_group->input_vals.push_back(inp);
+        continue;
+      }
+
+      // Could be something like a constant scalar, definition is nullptr, but
+      // isn't an "input" to the fusion. At least not one provided by an
+      // external source.
+      if (inp->definition() == nullptr) {
+        continue;
+      }
+
+      auto def_group = expr2group.at(inp->definition());
+      auto new_edge = segmented_fusion->newEdge(def_group, expr_group, inp);
+      expr_group->producer_edges.push_back(new_edge);
+      def_group->consumer_edges.push_back(new_edge);
+    }
+    for (auto out : expr->outputs()) {
+      if (out->isFusionOutput()) {
+        expr_group->output_vals.push_back(out);
+      }
+    }
+  }
+
+  bool merged_nodes = true;
+  while (merged_nodes) {
+    // Reset stateful traversal details in SegmentedGroups
+    resetTraversal();
+
+    resetLevels();
+
+    for (auto& group : groups()) {
+      if (group->merged) {
+        continue;
+      }
+      auto candidates = group->getMergeCandidates();
+      if (candidates.empty()) {
+        continue;
+      }
+
+      auto candidate_it = candidates.begin();
+      while (candidate_it != candidates.end() &&
+             !codeGenSupportedMerge(candidate_it->second)) {
+        candidate_it++;
+      }
+      if (candidate_it == candidates.end()) {
+        continue;
+      }
+
+      to_merge.emplace(group);
+      to_merge.emplace(candidate_it->first);
+
+      group->merged = true;
+      group->merge_with = candidate_it->first;
+      group->merge_through = candidate_it->second;
+
+      candidate_it->first->merged = true;
+      candidate_it->first->merge_with = group;
+      candidate_it->first->merge_through = candidate_it->second;
+    }
+
+    if (to_merge.empty()) {
+      merged_nodes = false;
+    }
+
+    mergeNodes();
+  }
```

Comment:
Would be really nice if we could have a generic DAG merger that could be used here as well as the expression sorting for loop-nest gen, but I guess that would be beyond the scope of this PR.

Reply:
I did start the semesters as being shared across expr sorting and segmentation, that didn't seem worthwhile, however I agree we should go through the logic in follow up PRs and see what we could effectively share.